plotbitrate.py

```diff
@@ -28,283 +29,596 @@
-    horizontalalignment='center', fontweight='bold', color='black')
-
-# calculate mean line position (right 85%, above line)
-mean_text_x = matplot.xlim()[1] * 0.85
-mean_text_y = global_mean_bitrate + \
-    ((matplot.ylim()[1] - matplot.ylim()[0]) * 0.015)
-mean_text = "mean ({:.0f})".format(global_mean_bitrate)
-
-# draw mean as think black line w/ text
-matplot.axhline(global_mean_bitrate, linewidth=2, color='black')
-matplot.text(mean_text_x, mean_text_y, mean_text,
-    horizontalalignment='center', fontweight='bold', color='black')
-
-matplot.legend()
-
-# render graph to file (if requested) or screen
-if args.output:
-    matplot.savefig(args.output, format=args.format)
-else:
-    matplot.show()
-
-# vim: ai et ts=4 sts=4 sw=4
+            values_min = sums_of_values
+            values_max = [
+                sum(pair) for pair in zip(sums_of_values, values)
+            ]
+        sums_of_values = values_max
+        color = Color[frame_type].value if frame_type in dir(Color) \
+            else Color.FRAME.value
+        bars[frame_type] = matplot.fill_between(
+            seconds, values_min, values_max, linewidth=0.5, color=color
+        )
+
+    return max(sums_of_values), int(statistics.mean(sums_of_values)), bars
+
+
+def add_area(
+        frames: Iterable[Frame],
+        duration: int,
+        downscale: bool,
+        max_display_values: int,
+        stream_type: str
+) -> Tuple[int, int]:
+    bitrates = OrderedDict(frames_to_kbits(frames, 0, duration))
+    bitrate_max = max(bitrates.values())
+    bitrate_mean = int(statistics.mean(bitrates.values()))
+
+    if downscale and 0 < max_display_values < duration:
+        factor = duration // max_display_values
+        bitrates = OrderedDict(downscale_bitrate(bitrates, factor))
+
+    seconds = list(bitrates.keys())
+    values = list(bitrates.values())
+    color = Color.AUDIO.value if stream_type == "audio" else Color.FRAME.value
+    matplot.plot(seconds, values, linewidth=0.5, color=color)
+    matplot.fill_between(seconds, 0, values, linewidth=0.5, color=color)
+    return bitrate_max, bitrate_mean
+
+
+def draw_horizontal_line_with_text(
+        pos_y: int,
+        pos_h_percent: float,
+        text: str
+) -> None:
+    # calculate line position (above line)
+    text_x = matplot.xlim()[1] * pos_h_percent
+    text_y = pos_y + ((matplot.ylim()[1] - matplot.ylim()[0]) * 0.015)
+
+    # draw as think black line with text
+    matplot.axhline(pos_y, linewidth=1.5, color="black")
+    matplot.text(
+        text_x, text_y, text,
+        horizontalalignment="center", fontweight="bold", color="black"
+    )
+
+
+def main():
+    args = parse_arguments()
+
+    # if the output is raw xml, just call the function and exit
+    if args.format == "xml_raw":
+        save_raw_xml(
+            args.input, args.output, args.stream_spec, args.no_progress
+        )
+        sys.exit(0)
+
+    duration = math.floor(media_duration(args.input))
```

Comment:
That's true. I also thought about making a version that works with streamable content.
That's one of the reasons I try to use frame iterables and avoid lists (until I really need them). I think this could be implemented in the future but there would be some other things to work on. Not just the media duration.

Reply:
Yeah this would be a fairly big rewrite of some parts. If that turns out to be requested by others or one of us find it useful we could add it then. Considering this a TODO for the future. Maybe your real time idea could solve this issue in a more elegant way.